src/modules/constructRelationalGraph.js

```diff
@@ -0,0 +1,265 @@
+        break;
+      default:
+        console.error('Invalid table name', table);
+        throw new Error('Invalid table name');
+    }
+  }
+  return columns.includes(attribute);
+};
+
+// function to check if a selection condition is a join condition
+// assuming that the selection condition is on a single table and join condition is between two tables
+const isValidJoinColumn = (attribute: string, tables: Array<string>) => {
+  if (attribute.includes('.')) {
+    const table = attribute.split('.')[0];
+    if (!tables.includes(table)) return false;
+  } else if (!conditionIsOnValidAttributes(attribute, tables)) {
+    return false;
+  }
+  return true;
+};
+
+export const getTableFromAttribute = (
+  attribute: string,
+  tables: Array<string>
+): string => {
+  if (attribute.includes('.')) {
+    return attribute.split('.')[0];
+  }
+  for (const table of tables) {
+    if (conditionIsOnValidAttributes(attribute, [table])) {
+      return table;
+    }
+  }
+  // this should never happen except for edge cases
+  return '';
+};
+
+const parseSelectionExpression = (
+  graph: Graph,
+  expr: Expression,
+  tables: Array<string>,
+  globalSelections: Array<SelectionCondition>
+) => {
+  if (ExpressionType.AND in expr) {
+    const conditions = expr[ExpressionType.AND].clauses;
+    for (const condition of conditions) {
+      parseSelectionExpression(graph, condition, tables, globalSelections);
+    }
+  } else if (ExpressionType.CMP in expr) {
+    const selection = expr[ExpressionType.CMP];
+    if (
+      isValidJoinColumn(selection.lhs, tables) &&
+      isValidJoinColumn(selection.rhs, tables)
+    ) {
+      const {lhs, rhs} = selection;
+      const leftTable = getTableFromAttribute(lhs, tables);
+      const rightTable = getTableFromAttribute(rhs, tables);
+      addEdge(graph, leftTable, rightTable, selection);
+    } else {
+      const {lhs, rhs} = selection;
+      const leftTable = getTableFromAttribute(lhs, tables);
+      const rightTable = getTableFromAttribute(rhs, tables);
+      const table = leftTable || rightTable;
+      if (table !== '') addNode(graph, table, expr);
+      else globalSelections.push(expr);
+    }
+  } else {
+    console.error(
+      'Invalid selection expression for Join Order Optimization',
+      expr
+    );
+  }
+};
+
+const parseJoinExpression = (
+  graph: Graph,
+  expr: Expression,
+  tables: Array<string>
+) => {
+  if (ExpressionType.AND in expr) {
+    const conditions = expr[ExpressionType.AND].clauses;
+    for (const condition of conditions) {
+      parseJoinExpression(graph, condition, tables);
+    }
+  } else if (ExpressionType.CMP in expr) {
+    const {lhs, rhs} = expr[ExpressionType.CMP];
+    const leftTable = getTableFromAttribute(lhs, tables);
+    const rightTable = getTableFromAttribute(rhs, tables);
+    addEdge(graph, leftTable, rightTable, expr[ExpressionType.CMP]);
+  } else {
+    console.error('Invalid join expression for Join Order Optimization', expr);
```

Comment:
What happens if this fails? You should make sure that the result is still a valid query even if you can't optimize it.

Reply:
This now throws an error, so rest of the optimization would not continue. The original/unoptimized query would be untouched.